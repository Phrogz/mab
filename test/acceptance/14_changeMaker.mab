setSlot( 'coins', Array new(1, 2) )
setSlot( 'countChange', method( amount,
	cc( amount, coins size, "-", 0 ) 
))

setSlot( 'cc', method( amount, kindsOfCoins, branch, depth, 
	p( "cc( #{amount}, #{kindsOfCoins}, #{branch}, #{depth} )" interpolate )
	if( amount ==(0), 
			1,
			if( amount <(0) or( kindsOfCoins ==(0) ), 
				0,
				cc( amount, kindsOfCoins -( 1 ), "a", depth +( 1 ) ) +( cc( amount -( coins at(kindsOfCoins) ), kindsOfCoins, "b", depth+( 1 ) ) )
			)
		)
	)
)

p( countChange( 3 ) )
