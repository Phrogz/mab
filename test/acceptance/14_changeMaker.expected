cc( 3, 2, -, 0 )
cc( 3, 1, a, 1 )
cc( 3, 0, a, 2 )
cc( 2, 1, b, 2 )
cc( 2, 0, a, 3 )
cc( 1, 1, b, 3 )
cc( 1, 0, a, 4 )
cc( 0, 1, b, 4 )
cc( 1, 2, b, 1 )
cc( 1, 1, a, 2 )
cc( 1, 0, a, 3 )
cc( 0, 1, b, 3 )
cc( -1, 2, b, 2 )
2